tools/utils.hh

```diff
@@ -61,4 +62,175 @@ const Op& get_selected_operation(int& ac, char**& av, const std::vector<Op>& ope
 // is configured with INFO level.
 void configure_tool_mode(app_template::seastar_options& opts, const sstring& logger_name = {});
 
+class basic_option {
+public:
+    const char* name;
+    const char* description;
+
+public:
+    basic_option(const char* name, const char* description) : name(name), description(description) { }
+
+    virtual void add_option(boost::program_options::options_description& opts) const = 0;
+};
+
+template <typename T = std::monostate>
+class typed_option : public basic_option {
+    std::optional<T> _default_value;
+    int _count;
+
+    virtual void add_option(boost::program_options::options_description& opts) const override {
+        if (_default_value) {
+            opts.add_options()(name, boost::program_options::value<T>()->default_value(*_default_value), description);
+        } else {
+            opts.add_options()(name, boost::program_options::value<T>(), description);
+        }
+    }
+
+public:
+    typed_option(const char* name, const char* description) : basic_option(name, description) { }
+    typed_option(const char* name, T default_value, const char* description) : basic_option(name, description), _default_value(std::move(default_value)) { }
+    typed_option(const char* name, const char* description, int count) : basic_option(name, description), _count(count) { }
+};
+
+template <>
+class typed_option<std::monostate> : public basic_option {
+    virtual void add_option(boost::program_options::options_description& opts) const override {
+        opts.add_options()(name, description);
+    }
+public:
+    typed_option(const char* name, const char* description) : basic_option(name, description) { }
+};
+
+class operation_option {
+    shared_ptr<basic_option> _opt; // need copy to support convenient range declaration of std::vector<option>
+
+public:
+    template <typename T>
+    operation_option(typed_option<T> opt) : _opt(make_shared<typed_option<T>>(std::move(opt))) { }
+
+    const char* name() const { return _opt->name; }
+    const char* description() const { return _opt->description; }
+    void add_option(boost::program_options::options_description& opts) const { _opt->add_option(opts); }
+};
+
+template <typename... Args>
+class operation {
+    using function = void(*)(Args...);
```

Comment:
How about std::function\<void ()>? then the function can hold the args. They should be constant, no?

It means we have to construct the operations map inside main() rather than statically in order to capture args, but it seems better than to template everything.

Reply:
> How about std::function<void ()>? then the function can hold the args. They should be constant, no?

The args are not constant. They are either derived from command line arguments or are results of common init code (e.g. creating a `sstables_manager`). All operations (commands) in a tool have common signature. To be able to pass down the work of selecting the operation according to the command selector argument, registering its command line parameters, generating the help msg, etc., to this central code, this code has to know the concrete operation type, which depends on the tool-specific signature.

Reply:
Maybe if I split the operator into the static part (name+description+args) and the function. I could declare the functions in `main` and they could capture all the args, while the static part could be declared statically. Would result in all operations being declared in two places.

Reply:
I have an idea on how to both avoid the template and keeping the single declaration of all oprations.

Reply:
Detemplated in v2.